### DISCUSSION TITLE
Connect to an external device using serial interface

### DISCUSSION THREAD

**Discussion by xostia** [arendst/Tasmota#22739] [General]
Hi,

I am developing a dark sky meter based on Tasmota and TSL2591 light sensor.

I would like to get the information (mpsas) from two commercial well known products: SQM-LU (serial connection) and Tess-W (wifi), this way, I can get the information of mine (FreeDSM), SQM-LU and Tess-W comparing performance.

With Tess-W is easy, I can connect from Berry to port 23 and get the information in JSON with tcpclient ().

SQM-LU only provides the information throught the USB port (TTL serial). Is there any way to connect to this serial port and get the information (I prepared the code to do it from the raspberry but It would be perfect to get data directly from Tasmota, this way I do not neet to add another device outdoor connected to the wifi).

Thank you for your support.

**Comment by barbudor** [comment]
No, you can't connect a USB port to an ESP
You may be able to hack the device to access it's internal serial if it is managed by a USB/serial converter. But if it isn't, you can only connect it to a USB host port of a computer /RPi

**Reply by barbudor** [reply]
You have to look into
- SQM-LR with rs232 although you will have to add a level converter from rs232 to TTL (such as Max232 or same family)
- SQM-LE with ethernet

**Comment by xostia** [comment]
Thank you for your fast answer... if I open the SQM and find the serial, the option is to use the Serial-to-TCP-Bridge?

**Reply by barbudor** [reply]
If you open it and can connect the SQM to the ESP through a True TTl-level serial, you will have different options:
- serial bridge to get the serial message through MQTT messages and optionaly process by rules. Depending on the protocol it may not be the simplest option
- TCP bridge: you will need some code of a remote (your RPI) to connect to the serial through the transparent serial bridge. Actually that would most likely convert your SQM to a SQM-LE (or a SQM-LW 😊) 
- Berry: using Berry's Serial, create a SQM bridge in a similar way as what you've done for the Tess-W (but using Serial instead of tcpclient)

**Reply by sfromis** [reply]
If you're more to C++ coding, an additional option for serial would be to create a native Tasmota driver using the existing `TasmotaSerial` driver. I'd still prefer Berry for handling it on an ESP32 with Tasmota.

**Reply by xostia** [reply]
Thank you, barbudor and sfromis... I will analyze the options.

One doubt... I was using 101 for my special sensor and driver for TSL2591 but now, in 14.4, it is used by other device.

I red somewhere in the Tasmota code that 91-99 are for user drivers ... but it is not true, now I can see all busy until 116.

Is there any user range to set my driver for TSL2591 and the future Tess-W and SQM ?

Thank you

**Reply by sfromis** [reply]
No, there is no reserved range for forking Tasmota to add nonstandard drivers. During development of drivers intended to be added to the project, you can pick some unused number, with the expectation of that being changed with the driver being integrated to the public project.

A nonstandard driver would basically have to be reintegrated and tested if wanting to do a new fork starting from a new Tasmota version, including *potential* updates of the driver to still work well with changed requirements. And driver number would not be the main consideration, just a detail.
